File: 0264-ugly-number-ii/0264-ugly-number-ii.java
class Solution {
    public int nthUglyNumber(int n) {
        int[] dp = new int[n+1];
        dp[1] = 1;
        int p2 = 1, p3 = 1, p5 = 1;
        for(int i = 2; i <= n; i++){
            int nextP2 = dp[p2]*2;
            int nextP3 = dp[p3]*3;
            int nextP5 = dp[p5]*5;
            dp[i] = Math.min(nextP2, Math.min(nextP3, nextP5));
            if(dp[i] == nextP2) p2++;
            if(dp[i] == nextP3) p3++;
            if(dp[i] == nextP5) p5++;
        }
        return dp[n];
    }
}

File: 0264-ugly-number-ii/0264-ugly-number-ii.ts
function nthUglyNumber(n: number): number {
    const dp: number[] = new Array(n+1).fill(0);
    dp[1] = 1;
    let p2: number = 1, p3:number = 1, p5:number = 1;
    for(let i: number = 2; i <= n; i++){
        let next2:number = dp[p2]*2;
        let next3:number = dp[p3]*3;
        let next5:number = dp[p5]*5;
        dp[i] = Math.min(next2,next3,next5);
        if(dp[i] == next2) p2++;
        if(dp[i] == next3) p3++;
        if(dp[i] == next5) p5++;
    }
    return dp[n];
};